### PR TITLE
feat: implement sdk for session reduce

### PR DESCRIPTION
**Pull request by KeranYang**
Tested using the existing golang sdk e2e: https://github.com/KeranYang/numaflow/pull/98

### CODE REVIEW

**Comment by whynowy** [comment]
Convert to draft mode since it's taking too long to review.

File: src/main/java/io/numaproj/numaflow/sessionreducer/SessionReducerActor.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        if (this.isClosed) {
+            throw new RuntimeException(
+                    "received an expand request but the session is already closed.");
+        }
+        if (this.isMerging) {
+            throw new RuntimeException(
+                    "cannot expand a session window when it's in a merging process."
+                            + " window info: " + this.keyedWindow.toString());
+        }
+        // update the keyed window
```
Comment (yhl25) [inline src/main/java/io/numaproj/numaflow/sessionreducer/SessionReducerActor.java:76]: it is valid to have an `expand` operation during merging right?

File: src/main/java/io/numaproj/numaflow/sessionreducer/SessionReducerActor.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    private void invokeHandler(HandlerDatum handlerDatum) {
+        if (this.isClosed) {
+            throw new RuntimeException(
+                    "received a message but the session is already closed.");
+        }
+        if (this.isMerging) {
+            throw new RuntimeException(
+                    "cannot process messages when the session window is in a merging process."
+                            + " window info: " + this.keyedWindow.toString());
+        }
```
[the remark this answers is not in the document]
Comment (KeranYang) [reply src/main/java/io/numaproj/numaflow/sessionreducer/SessionReducerActor.java]: Removed.

File: src/main/java/io/numaproj/numaflow/sessionreducer/Service.java [hunk 3]
```diff
@@ -79,7 +84,29 @@ public StreamObserver<Sessionreduce.SessionReduceRequest> sessionReduceFn(final
             public void onNext(Sessionreduce.SessionReduceRequest sessionReduceRequest) {
                 // send the message to parent actor, which takes care of distribution.
                 if (!supervisorActor.isTerminated()) {
-                    supervisorActor.tell(sessionReduceRequest, ActorRef.noSender());
+                    // if the operation is a MERGE, make it a blocking call.
+                    if (sessionReduceRequest.getOperation().getEvent()
+                            == Sessionreduce.SessionReduceRequest.WindowOperation.Event.MERGE) {
+                        // set time out to 1 second as we expect a MERGE operation to finish quickly.
+                        Timeout timeout = new Timeout(Duration.create(1, "seconds"));
+                        try {
```
Comment (yhl25) [inline src/main/java/io/numaproj/numaflow/sessionreducer/Service.java:92]: Hacky, sometimes it might take more than 1 second

Comment (KeranYang) [reply src/main/java/io/numaproj/numaflow/sessionreducer/Service.java:92]: thanks, what should be the proper number? @yhl25

Comment (yhl25) [reply src/main/java/io/numaproj/numaflow/sessionreducer/Service.java:92]: let's wait until the it returns

Comment (yhl25) [reply src/main/java/io/numaproj/numaflow/sessionreducer/Service.java:92]: also, please add unit tests to cover merge followed by another operation to make sure the blocking call works!

Comment (KeranYang) [reply src/main/java/io/numaproj/numaflow/sessionreducer/Service.java:92]: The method Await asks for a timeout, I don't think we should wait forever, right? Maybe we should set an atMost timeout value.

Yes, the uts for merge are already in place. Please check the serverTest class.

**Review by yhl25 — APPROVED** [review]

**Pull request merged**